src/companion.jl

```diff
@@ -15,86 +16,78 @@ Also, directly from a polynomial:
 ```julia
 julia> using Polynomials
 
-julia> P=Polynomial([2.0,3,4,5])
+julia> P = Polynomial(2:5)
 Polynomial(2 + 3x + 4x^2 + 5x^3)
 
-julia> C=Companion(P)
+julia> C = Companion(P)
 3×3 Companion{Float64}:
  0.0  0.0  -0.4
  1.0  0.0  -0.6
  0.0  1.0  -0.8
 ```
 """
-struct Companion{T} <: AbstractArray{T, 2}
+struct Companion{T} <: AbstractMatrix{T}
     c :: Vector{T}
 end
 
-# Generate companion matrix from a polynomial
+Companion(v::AbstractVector) = Companion(collect(v))
 
-function Companion(P::Polynomial{T}) where T
-   n = length(P)
-   c = Array{T}(undef, n-1)
-   d=P.coeffs[n]
-   for i=1:n-1
-       c[i]=P.coeffs[i]/d
-   end
-   Companion(c)
+# Construct companion matrix from a polynomial
+
+function Companion(P::Polynomial)
+    n = length(P)
+    c = P.coeffs[1:(n-1)] / P.coeffs[n]
+    return Companion(c)
 end
 
-#Basic property computations
-size(C::Companion, r::Int) = (r==1 || r==2) ? length(C.c) :
-    throw(ArgumentError("Companion is of rank 2"))
+# Basic properties
 
 function size(C::Companion)
     n = length(C.c)
-    n, n
+    return n, n
 end
 
-#XXX Inefficient but works
-# getindex(C::Companion, i, j) = getindex(Matrix(C), i, j)
-# isassigned(C::Companion, i, j) = isassigned(Matrix(C), i, j)
-getindex(C::Companion{T}, i::Int, j::Int) where T = (j==length(C.c)) ? -C.c[i] : (i==j+1 ? one(T) : zero(T) )
-isassigned(C::Companion, i::Int, j::Int) = (j==length(C.c)) ? isassigned(C.c,i) : true
+@inline Base.@propagate_inbounds function getindex(
+    C::Companion{T},
+    i::Int,
+    j::Int,
+) where T
+    @boundscheck checkbounds(C, i, j)
+    return (j == length(C.c)) ? (@inbounds -C.c[i]) :
+        i == j+1 ? one(T) : zero(T)
+end
 
 
-function Matrix(C::Companion{T}) where T
-    M = zeros(T, size(C)...)
-    M[:,end]=-C.c
-    for i=1:size(C,1)-1
-        M[i+1, i] = one(T)
-    end
-    M
-end
+# Linear algebra
 
-#Linear algebra stuff
-function mul!(C::Companion{T}, b::Vector{T}) where T
-    x = b[end]
-    y = -C.c[1]*x
-    b[2:end] = b[1:end-1]-C.c[2:end]*x
-    b[1] = y
-    b
+# 3-argument mul! mutates first argument: y <= C * x
+function mul!(y::Vector, C::Companion, x::AbstractVector)
+    @boundscheck length(y) == length(x) == size(C, 1) ||
+        throw(DimensionMismatch("mul! arguments incompatible sizes"))
+    z = x[end]
+    y[1] = -C.c[1] * z
+    y[2:end] = x[1:(end-1)] - C.c[2:end] * z
```

Comment:
```suggestion
    y[2:end] = x[begin:end-1] - C.c[2:end] * z
```

Reply:
With `x` and `AbstractVector` there are no guarantees that indexing starts with `1`.